text/0000-otlp-over-messaging-systems.md

```diff
@@ -0,0 +1,110 @@
+# OTLP over messaging systems
+
+Best practices for transporting OTLP data over a message system.
+
+## Motivation
+
+This proposal tries to bring consistency and guidelines when transporting
+OTLP data over messaging systems. A non-exclusive list of examples
+of products in this category are:
+
+* Apache Kafka
+* Apache Pulsar
+* Google Pubsub
+* AWS SNS
+* RabbitMQ
+
+Using an intermediate messaging system is in some cases preferred compared to
+a direct network connection between the telemetry producer and consumer.
+Reasons for using such an intermediate medium could be:
+
+* Protecting against backend failure
+* Security Policies
+* Network Policies
+* Buffering
+
+An extra motivation to have a consistent definition of the payload is that
+it would be easy to transfer the OTLP data from one messaging system to
+another just by reading the payload from one system and writing it to
+another without the need for transformations.
+
+## Explanation
+
+Because the OTLP payload that is sent over messaging systems is
+well-defined, it’s easy to implement new systems consistently.
+An implementation should at least support `otlp_proto`, meaning that the
+payload is Protobuf serialized from `ExportTraceServiceRequest` for traces,
+`ExportMetricsServiceRequest` for metrics, or `ExportLogsServiceRequest` for
+logs.
+
+Optionally an implementation could support `otlp_json` or other alternative
+encodings like jaeger or zipkin payloads. Alternative encodings should be configured
+by an `encoding` field in the configuration file.
+
+For the user it’s beneficial that both an exporter, and a receiver are
+implemented in the OpenTelemetry Collector, but SDK developers could
+also implement a language specific exporter.
+
+For log data, implementors of a receiver are encouraged to also support
+receiving raw data from a topic. This enables scenarios were
+non-OpenTelemetry log producers can produce a stream of logs, either
+structured or unstructured, on that topic and the receiver wraps the log
+data in a configurable resource.
+
+## Internal details
+
+The default implementation must implement `otlp_proto`, meaning that the
+payload is Protobuf serialized from `ExportTraceServiceRequest` for traces,
+`ExportMetricsServiceRequest` for metrics, or `ExportLogsServiceRequest` for
```

Comment:
One thing that is not clear to me is if we want to include the signal type in the serialized payload so that he recipient can tell what signal it is. Since this OTEP does not define it it means the recipient that works with multiple signals needs another way to figure it out (often ends up using the channel name). I think this is an important capability that adds value and has no obvious downsides.

This is also an issue we want to solve for OTLP/File format: https://github.com/open-telemetry/opentelemetry-specification/issues/1443#issuecomment-780183624
It will be good to solve this here as well and do it in the same way we will do for OTLP/File so this likely requires to think through the OTLP/File design as well.

Reply:
This doesn't help with the file format at all, but Kafka and Pubsub (maybe others) have key-value headers/attributes which are mentioned in "Open questions". The signal type and encoding (json or proto) could go in there

Reply:
I've added a comment in the OTLP/File discussion about CloudEvents. Certainly, if you want to allow multiplexing logs, traces, and metrics over a single topic CloudEvents is a good fit and it would align with this OTLP as well as the spec defines bindings for different messaging systems:

- [Kafka](https://github.com/cloudevents/spec/blob/v1.0.1/kafka-protocol-binding.md)
- [MQTT](https://github.com/cloudevents/spec/blob/v1.0.1/mqtt-protocol-binding.md)
- [NATS](https://github.com/cloudevents/spec/blob/v1.0.1/nats-protocol-binding.md)
- [Google Pubsub](https://github.com/google/knative-gcp/blob/main/docs/spec/pubsub-protocol-binding.md)

Reply:
@alexvanboxel I read the comment you left in the OTLP/File issue but I am a bit confused why it talks about the CloudEvents or Avro.

My comment here is about how an OTLP payload that is encoded as a Protobuf or JSON should be recorded in messaging systems such that the signal type information is included. I think we should keep it as simple as possible, new any additional concepts and technology introduced (CloudEvents or Avro) IMO is unnecessary complication.

If we want to record the type of the signal there are fairly ways to achieve it while staying with the existing set of technologies. For example we can add this message to OTLP Protobuf definitions:
```protobuf
// This may be also named simply Request or Payload or something like that
message ExportServiceRequest {
  oneof request {
    ExportTraceServiceRequest traces = 1;
    ExportMetricsServiceRequest metrics = 2;
    ExportLogsServiceRequest logs = 3;
  }
}
```
ExportServiceRequest can be then serialized and become the payload in the messaging system. Recipients can then deserialize and will be able to use the right signal in the "request" field.

Reply:
@tigrannajaryan the consensus is that we want to signal the type of the message, so I'm already changing my application. Let me demonstrate what I mean by leveraging the CloudEvent spec (as it's a spec and not technology). In my Google Pubsub implementation, it describes what headers to use. So I could simple use:

```
func (receiver *pubsubReceiver) detectEncoding(attributes map[string]string) (Encoding, error) {
	ceType := attributes["ce-type"]
	ceContentType := attributes["ce-datacontenttype"]
	if strings.HasSuffix(ceContentType,"application/x-protobuf") {
		switch ceType {
		case "org.opentelemetry.otlp.traces.v1":
			return OtlpProtoTrace, nil
		case "org.opentelemetry.otlp.metrics.v1":
			return OtlpProtoMetric, nil
		case "org.opentelemetry.otlp.logs.v1":
			return OtlpProtoLog, nil
		}
	} else if strings.HasSuffix(ceContentType,"application/json") {
		switch ceType {
		case "org.opentelemetry.otlp.traces.v1":
			return OtlpJsonTrace, nil
		case "org.opentelemetry.otlp.metrics.v1":
			return OtlpJsonMetric, nil
		case "org.opentelemetry.otlp.logs.v1":
			return OtlpJsonLog, nil
		}
	}
	return Unknown, nil
}
```

Granted, it's heavier than a `oneOf`, but it leaves the flexibility of even signaling OTLP/Json encoding.